File: src/main/java/fr/arsenelapostolet/professor/server/core/application/GradesApplication.java
package fr.arsenelapostolet.professor.server.core.application;

import com.google.inject.Inject;
import fr.arsenelapostolet.professor.server.core.contracts.CreateClassResponse;
import fr.arsenelapostolet.professor.server.core.entities.Student;
import fr.arsenelapostolet.professor.server.core.exceptions.InvalidResourceException;
import fr.arsenelapostolet.professor.server.core.repositories.StudentRepository;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.HashSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class GradesApplication {

    private final StudentRepository studentRepository;

    @Inject
    public GradesApplication(StudentRepository studentRepository) {
        this.studentRepository = studentRepository;
    }

    public CreateClassResponse createClass(String className, Stream<String> csvLines) {
        final var students = csvLines
                .skip(1)
                .map(line -> line.split(","))
                .map(csvLine -> parseStudentFromCsv(className, csvLine));

        return new CreateClassResponse(
                className,
                studentRepository.saveStudents(students.collect(Collectors.toSet()))
        );
    }

    private Student parseStudentFromCsv(String className, String[] csvLine) {
        try {
            return new Student(
                    csvLine[0],
                    csvLine[1],
                    csvLine[2],
                    csvLine[4],
                    csvLine[5],
                    new HashSet<>(),
                    className,
                    new URI(csvLine[6])
            );
        } catch (URISyntaxException uriSyntaxException) {
            throw new InvalidResourceException(
                    String.format(
                            "Student with id %s has an invalid project URL : %s", csvLine[0],
                            csvLine[6]
                    ),
                    uriSyntaxException
            );
        }
    }

}

File: app/src/main/kotlin/fr/arsenelapostolet/professor/core/application/GradesApplication.kt
package fr.arsenelapostolet.professor.core.application

import fr.arsenelapostolet.professor.core.entities.Student
import fr.arsenelapostolet.professor.core.exceptions.InvalidResourceException
import fr.arsenelapostolet.professor.core.repositories.StudentRepository
import java.net.URI
import java.net.URISyntaxException

class GradesApplication(private val studentRepository: StudentRepository) {

    fun getStudent(): Collection<Student> {
        return studentRepository.getAllStudents()
    }

    fun importStudents(csvLines: Collection<String>): Set<Student> {
        val students = csvLines
            .drop(1)
            .map { line: String -> line.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray() }
            .map { csvLine: Array<String> -> parseStudentFromCsv(csvLine) }

        return studentRepository.saveStudents(students.toSet())
    }

    private fun parseStudentFromCsv(csvLine: Array<String>): Student {
        if(csvLine.size < 7){
            throw InvalidResourceException(
                String.format(
                    "Line with id %s too few columns : %s",
                    csvLine[0],
                    csvLine.joinToString(",")
                )
            )
        }

        try {
            return Student(
                csvLine[1],
                csvLine[2],
                csvLine[3],
                csvLine[5],
                csvLine[6],
                mutableListOf(),
                csvLine[0],
                URI(csvLine[7])
            )
        } catch (uriSyntaxException: URISyntaxException) {
            throw InvalidResourceException(
                String.format(
                    "Student with id %s has an invalid project URL : %s", csvLine[0],
                    csvLine[6]
                ),
                uriSyntaxException
            )
        }
    }
}
